### PR TITLE
FEATURE: adding explicit method for black scholes (using forked repo)

### PR DESCRIPTION
**Pull request by chemardes**
The following implementation only takes into account a european call option

Changes:
- Implemented the explicit method on the black-scholes equation

Things to do:

- Need to implement it to allow toggling between call/put options (user can set it in the input parameters? e.g. "call"/"put")
- Need to add check to see if user-defined conditions fall within the CFL condition
- Need to add parameter validation (e.g. nodes need to be positive integers)
- Error handling
- Try to implement it in Crank Nicolson
- Add comments

### CODE REVIEW

File: playground/black_scholes_explicit.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import numpy as np
+from matplotlib import pyplot as plt
+
+
+class BlackScholesExplicitSolver():
```
Comment (alphaville) [inline playground/black_scholes_explicit.py:5]: Can you just add a couple of lines of documentation? Include the problem statement together with the initial/boundary conditions.

File: playground/black_scholes_explicit.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        return q
+
+    def __generate_asset_grid(self):
+        return np.linspace(0, self.__S_max, self.__s_nodes + 1)
+
+    def __generate_time_grid(self):
+        return np.linspace(0, self.__expiry, self.__t_nodes + 1)
+
+    def __set_boundary_conditions(self, q, T, tau):
```
Comment (chemardes) [inline playground/black_scholes_explicit.py:119]: is there a better way to do this?

File: playground/black_scholes_cn.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        V[0, :] = 0
+        V[-1, :] = S[-1] - self.__K * np.exp(-self.__r * (self.__expiry - T))
+
+        for tau in reversed(range(self.__t_nodes)):
+            # Construct the RHS vector for this time step
+            rhs_vector = rhs @ V[1:-1, tau + 1]
+
+            # Apply boundary conditions to the RHS vector
+            rhs_vector[0] = alpha[1] * (V[0, tau + 1] + V[0, tau]) - alpha[0] * V[0, tau+1]
+            rhs_vector[-1] = gamma[self.__s_nodes-1] *(V[-1, tau+1] + V[-1, tau])
```
Comment (chemardes) [inline playground/black_scholes_cn.py:62]: there might be something wrong with the way im doing this - my plot looks weird

Screenshot 2024-11-27 at 19 08 28

Comment (chemardes) [reply playground/black_scholes_cn.py]: fixed with latest commit

**Comment by chemardes** [comment]
@alphaville @debdal6 free to merge?

**Pull request merged**